server/schedule/labeler/rules.go

```diff
@@ -68,20 +73,79 @@ type LabelRulePatch struct {
 	DeleteRules []string     `json:"deletes"`
 }
 
+func (l *RegionLabel) expireBefore(t time.Time) bool {
+	if l.expire == nil {
+		return false
+	}
+	return l.expire.Before(t)
+}
+
+func (l *RegionLabel) checkAndAdjustExpire() (err error) {
+	if len(l.TTL) == 0 {
+		l.expire = nil
+		return
+	}
+	ttl, err := time.ParseDuration(l.TTL)
+	if err != nil {
+		return err
+	}
+	var startAt time.Time
+	if len(l.StartAt) == 0 {
+		startAt = time.Now()
+		l.StartAt = startAt.Format(time.UnixDate)
+	} else {
+		startAt, err = time.Parse(time.UnixDate, l.StartAt)
+		if err != nil {
+			return err
+		}
+	}
+	expire := startAt.Add(ttl)
+	l.expire = &expire
+	return nil
+}
+
+func (rule *LabelRule) checkAndRemoveExpireLabels(now time.Time) bool {
+	labels := make([]RegionLabel, 0)
+	rule.minExpire = nil
+	for _, l := range rule.Labels {
+		if l.expireBefore(now) {
+			continue
+		}
+		labels = append(labels, l)
+		if rule.minExpire == nil || l.expireBefore(*rule.minExpire) {
+			rule.minExpire = l.expire
+		}
+	}
+
+	if len(labels) == len(rule.Labels) {
+		return false
+	}
+	rule.Labels = labels
+	return true
+}
+
 func (rule *LabelRule) checkAndAdjust() error {
 	if rule.ID == "" {
 		return errs.ErrRegionRuleContent.FastGenByArgs("empty rule id")
 	}
 	if len(rule.Labels) == 0 {
 		return errs.ErrRegionRuleContent.FastGenByArgs("no region labels")
 	}
-	for _, l := range rule.Labels {
+	for id, l := range rule.Labels {
 		if l.Key == "" {
 			return errs.ErrRegionRuleContent.FastGenByArgs("empty region label key")
 		}
 		if l.Value == "" {
 			return errs.ErrRegionRuleContent.FastGenByArgs("empty region label value")
 		}
+		if err := rule.Labels[id].checkAndAdjustExpire(); err != nil {
```

Comment:
```suggestion
		if err := l.checkAndAdjustExpire(); err != nil {
```

Reply:
Here we should use `rule.Labels[id]` instead of `l` because `l` is a *copy*(not the address) of the element `rule.Labels[id]`,  and we need to adjust the value of `rule.Labels[i]` here.

Reply:
got.